File: src/eclair_web.c
/*
 * eclair - Web backend: ariaNotify (screen reader), Web Speech (synthesizer)
 * Copyright (c) 2026 Jesse Jurman. zlib license - see LICENSE.md
 */

#include "eclair_backend.h"

#if defined(__EMSCRIPTEN__)

// functions we want exposed and available in eclair_web.js
extern bool eclair_js_sr_speak(const char *utf8, bool interrupt);
extern bool eclair_js_synth_available(void);
extern bool eclair_js_synth_speak(const char *utf8, bool interrupt);
extern bool eclair_js_synth_stop(void);
extern void eclair_js_synth_set_rate(float rate);
extern void eclair_js_synth_set_volume(float volume);

// SpeechSynthesisUtterance.rate defaults at 1.0 and has a range of 0.1 to 10
// to provide a sane range comparable to other backends, we set the max to 3.0
#define ECLAIR_WEB_RATE_MIN 0.5f
#define ECLAIR_WEB_RATE_MID 1.0f
#define ECLAIR_WEB_RATE_MAX 3.0f

/* ---------------------------------------------------------------
 * lifecycle
 * --------------------------------------------------------------- */

bool eclair_platform_init(void) {
	// nothing to do here, always return true
	return true;
}

void eclair_platform_shutdown(void) {
	// nothing to do here
}

/* ---------------------------------------------------------------
 * screen reader route - ariaNotify
 * --------------------------------------------------------------- */

bool eclair_sr_available(void) {
	// we can't determine from the browser if a screen reader is running, so by
	// default we assume one could be running.
	return true;
}

bool eclair_sr_speak(const char *utf8, bool interrupt) {
	return eclair_js_sr_speak(utf8, interrupt);
}

bool eclair_sr_stop(void) {
	// no-op
	return true;
}

const char *eclair_sr_name(void) {
	return "ariaNotify";
}

/* ---------------------------------------------------------------
 * synthesizer route - Web Speech
 * --------------------------------------------------------------- */

bool eclair_synth_available(void) {
	return eclair_js_synth_available();
}

bool eclair_synth_speak(const char *utf8, bool interrupt) {
	return eclair_js_synth_speak(utf8, interrupt);
}

bool eclair_synth_stop(void) {
	return eclair_js_synth_stop();
}

void eclair_synth_set_rate(float rate) {
	eclair_js_synth_set_rate(
		eclair_map_rate(rate, ECLAIR_WEB_RATE_MIN, ECLAIR_WEB_RATE_MID, ECLAIR_WEB_RATE_MAX)
	);
}

void eclair_synth_set_volume(float volume) {
	eclair_js_synth_set_volume(volume);
}

const char *eclair_synth_name(void) {
	return "Web Speech";
}

#else
// Not an emscripten platform
#endif /* __EMSCRIPTEN__ */

File: src/eclair_web.js
/*
 * eclair - Web backend implementation
 * Copyright (c) 2026 Jesse Jurman. zlib license - see LICENSE.md
 *
 * Every function returns 1 or 0 rather than true or false. A C bool arrives as
 * an i32, and returning a JS boolean leaves the coercion to chance.
 */

mergeInto(LibraryManager.library, {
	$eclairWebState: {
		rate: 1.0,
		volume: 1.0
	},

	/* ------------------------------------------------- *
	 * screen reader route - ariaNotify
	 * ------------------------------------------------- */

	eclair_js_sr_speak__deps: ['$UTF8ToString'],
	eclair_js_sr_speak: function (utf8, interrupt) {
		try {
			document.ariaNotify(UTF8ToString(utf8), {
				priority: interrupt ? "high" : "normal"
			});
			return 1;
		} catch (e) {
			// note - we can end up here when we have no document (e.g. web-worker)
			return 0;
		}
	},

	/* ------------------------------------------------- *
	 * synthesizer route - Web Speech
	 * ------------------------------------------------- */

	eclair_js_synth_available: function () {
		return (typeof speechSynthesis !== "undefined") ? 1 : 0;
	},

	eclair_js_synth_speak__deps: ['$UTF8ToString', '$eclairWebState'],
	eclair_js_synth_speak: function (utf8, interrupt) {
		if (typeof speechSynthesis === "undefined") {
			return 0;
		}

		try {
			if (interrupt) {
				speechSynthesis.cancel();
			}

			var utterance = new SpeechSynthesisUtterance(UTF8ToString(utf8));
			utterance.rate = eclairWebState.rate;
			utterance.volume = eclairWebState.volume;

			speechSynthesis.speak(utterance);
			return 1;
		} catch (e) {
			return 0;
		}
	},

	eclair_js_synth_stop: function () {
		if (typeof speechSynthesis === "undefined") {
			return 0;
		}

		speechSynthesis.cancel();
		return 1;
	},

	eclair_js_synth_set_rate__deps: ['$eclairWebState'],
	eclair_js_synth_set_rate: function (rate) {
		eclairWebState.rate = rate;
	},

	eclair_js_synth_set_volume__deps: ['$eclairWebState'],
	eclair_js_synth_set_volume: function (volume) {
		eclairWebState.volume = volume;
	}

})
